### ISSUE TITLE
LRT of the main effect of a continuous variable is always 0 in the output of clusterperm.lmer

### ISSUE THREAD

**Issue by hy-lu** [cvoeten/permutes#4]
Hi there, 
I've recently found another bug probably related to the last one I issued #3. In the output of my analyses, the LRT of a continuous variable is always 0. I guess this is the result of the following code:
https://github.com/cvoeten/permutes/blob/51351d35e8d5f88fe651c23758aea85e690f2f20/R/buildmer.R#L227
When constructing the restricted model, the removal of the main effect of a factor variable with the presence of its higher-order interaction results in the same number of columns in the model matrix. Here is a minimal example:
```{r}
set.seed(1)
toy_data <- expand.grid(A = c("A1", "A2"), B = rnorm(10, 5, 1)) %>% add_column(Y = rnorm(20, 10, 5))
lm(Y ~ 1 + B + A:B, data = toy_data) %>% model.matrix() %>% head()
>   (Intercept)        B      B:A1
> 1           1 4.373546  4.373546
> 2           1 4.373546 -4.373546
> 3           1 5.183643  5.183643
> 4           1 5.183643 -5.183643
> 5           1 4.164371  4.164371
> 6           1 4.164371 -4.164371

lm(Y ~ 1 + A + A:B, data = toy_data) %>% model.matrix() %>% head()
>   (Intercept) A1    AA1:B    AA2:B
> 1           1  1 4.373546 0.000000
> 2           1 -1 0.000000 4.373546
> 3           1  1 5.183643 0.000000
> 4           1 -1 0.000000 5.183643
> 5           1  1 4.164371 0.000000
> 6           1 -1 0.000000 4.164371
```
Thus, the function would return 0 for LRT and df in this scenario. I'm not sure what the `if` statement here is for but I guess a workaround is to change it to `if (!(NCOL(X.full) >= NCOL(X.restricted)))`.

----

There are some other bugs that I've also found. One is https://github.com/cvoeten/permutes/blob/51351d35e8d5f88fe651c23758aea85e690f2f20/R/buildmer.R#L216, I think it should be `			X[, colnames(X) != '(Intercept)'] <- 0` (A comma is missed).

The other is https://github.com/cvoeten/permutes/blob/51351d35e8d5f88fe651c23758aea85e690f2f20/R/buildmer.R#L295 https://github.com/cvoeten/permutes/blob/51351d35e8d5f88fe651c23758aea85e690f2f20/R/buildmer.R#L346

In some of my analyses, column `t` & `p` were not matched/aligned with `Factor`, `LRT` and `beta`, such as the example as follows. The output on the left is returned by `afex::mixed`, and the output on the right is returned by `permutes::perm.lmer`.
![c3d5c3cac8ec08fb72026e3d1c7e494](https://user-images.githubusercontent.com/8774247/141754762-2e47352b-6610-4fa3-a7b4-49d42ddf7dc2.png)

Instead of putting them in a data frame by the "natural" row order, I think it would be more robust when creating the data frame is matched by the names of each vector.

**Comment by cvoeten** [comment]
Thanks very much for your detailed report. I need some time to think about the best solution. Here are some notes, for now:

1. The early return is meant as an optimization, in case the term we're testing does not actually affect the model fit. It's actually meant for a similar case to yours: suppose I have a model Y ~ 1 + A and I want to test the intercept, so I want to compare that to Y ~ 0 + A. Well, those two models are actually equivalent (not in terms of coefficients, but they are in log-likelihoods and hence in LRT statistics), because the first model has two columns, one for the intercept and one for the contrast A versus B, and the second model has two columns, one for A and one for B. Since these will by definition have the same likelihoods, no point in doing the whole permutation dance here.
Note that the case you're presenting here is more complicated because your two models ARE genuinely different... so this whole check needs to be changed. Maybe I should just remove it, as the permutations will naturally become zero anyway.

2. That's not a bug - I was thinking this might be more robust in the general case, should permutes ever be extended to packages which might drop dims in case you'd end up with a 1-column or 1-row model matrix. But I'll change it anyway - I agree it looks suspicious enough to cause confusion (as you've demonstrated to me) and it isn't necessary for lm or lmer as they behave nicely even for these edge cases.

3. This is really, really, really bad, thanks for pointing it out to me!! I need to think on how to solve this - your solution of indexing the dataframe by the real rownames seems clever, but will not work, because we don't have those names for the permutes-computed LRT statistics and p-values. I have no idea how to solve this at the moment, and need to give this a lot more thought (how are these even getting jumbled in the first place?) - I fear I might actually need to do a total rework of my approach to isolating model terms, ditching buildmer's terms approach and working exclusively with the model matrix (but then how would ANOVA work?). Suggestions - particularly for less drastic approaches - very welcome!

**Comment by hy-lu** [comment]
Thanks for the timely response! 🙇 

1. Indeed, I haven't thought of that in the case you mentioned these models are equivalent. But I agree that removing the check can help fix this problem only at the expense of computation time.

2. I think when the dim of X matrix is dropped, the function `colnames` would return NULL, which brings out other issues. I suppose that `X[colnames(X) != "(Intercept)"] <- 0` tries to isolate only the effect of Intercept. However, in my case, it didn't work as expected because it changed the matrix elements to zero by row:
```{r}
Browse[4]> head(X)
  (Intercept) response_response1   aq_tot aq_tot:response1
1           1                 -1 1.555174        -1.555174
2           1                  1 1.555174         1.555174
3           1                 -1 1.555174        -1.555174
4           1                  1 1.555174         1.555174
5           1                 -1 1.555174        -1.555174
6           1                 -1 1.555174        -1.555174
Browse[4]> X[colnames(X) != "(Intercept)"] <- 0
Browse[4]> head(X)
  (Intercept) response_response1   aq_tot aq_tot:response1
1           1                  0 0.000000         0.000000
2           0                  1 0.000000         0.000000
3           0                  0 1.555174         0.000000
4           0                  0 0.000000         1.555174
5           1                  0 0.000000         0.000000
6           0                 -1 0.000000         0.000000
```

3. When I'm probing the problem, I found that the function would throw errors as follows when a factor have more than two levels (in the following case, the discrete variable has three levels):
```{r}
Error in `$<-.data.frame`(`*tmp*`, cluster, value = NA) : 
  replacement has 1 row, data has 0
In addition: Warning message:
In data.frame(Factor = unname(terms), LRT = unname(LRT), beta = unname(beta),  :
  arguments imply differing number of rows: 5, 6
```
The error seems stem from the "term naming" strategy from Lines 154-169 where only the main effect of a factor would be processed but the interaction terms were unchanged. Eventually, LRT would have fewer terms than those returned by buildmer since the interaction would be separated into more terms in the regression table by lmer when factors have more than two levels. But the type "anova" works pretty well for such case.
I'm thinking if it's possible to make new columns/terms for interactions just like what you did for factor variables in Lines 154-169 and use the new interaction terms in the buildmer so that you can decide how the terms are named in buildmer. 
> ... because we don't have those names for the permutes-computed LRT statistics and p-values.
Also, I find the vector of permutes-computed LRT statistics for the output data frame is actually named so I'm not sure where this were referring.

Hope this could help you somehow! 🙏

**Comment by cvoeten** [comment]
First of all, thanks for your extensive digging and your patience. I've now significantly reworked the affected code. Instead of trying to identify model-matrix names and trying to shuffle things into place after they get screwed up, I now simply internally construct my own data with the very appealing names X1, X2, X3, ... and I restore them when the model has been fitted. This simplifies quite a lot of the code, and hopefully has now also managed to solve all of the problems that you've identified. Can you test the current Github version?

**Comment by cvoeten** [comment]
I've made some further fixes to this new approach (e.g. ANOVA now works) and as far as I can see, everything now appears to be correct. This is already on GitHub and will become available generally in permutes v2.3. I'll close this issue now, but please reopen if you find that your dataset still demonstrate a problem.

**Comment by hy-lu** [comment]
Hi, I tried the most recent version, and it will not throw errors anymore when the cluster-related terms are NA. That's perfect!

But I'm wondering if it's the best way to forcibly set the intercept to NA as in
https://github.com/cvoeten/permutes/blob/b27c026a6e03e90a82aaa8b6380c287004d8bccd/R/buildmer.R#L240

First, I think this resulted in the NA perms for the intercept https://github.com/cvoeten/permutes/issues/4#issuecomment-992248520 in the first place, and this would happen not only for the intercept-only model but for any model with intercept.

Second, my major concern is that the intercept term still makes sense and is worth being included, especially when using orthogonal predictors it represents the grand mean. Moreover, the cluster-related columns of the intercept in the final output df will also be NA for type = "regression".

What do you think?

**Comment by cvoeten** [comment]
OK, glad that it works now!

I believe setting the intercept to NA is correct. The approach to permutation testing that is used in these functions is to obtain the marginal errors (i.e. the residuals plus the effect of interest), and then comparing a 'true' model to a permuted model. The 'true' model predicts these marginal errors from the predictor of interest; if the predictor does something that has a higher signal than the noise from the errors, then we will get a certain likelihood-ratio statistic from that model. The permuted model randomly shuffles the predictor and then computes the same log-likelihood - this is done a large number of times to get a distribution of likelihood-ratio statistics in which the predictor has been randomly shuffled. The assumption is that this random shuffling reflects a null hypothesis, such that if the random shuffles achieve lower likelihood ratios than the 'true' model, the 'true' model will be assigned a low p-value.

However, randomly shuffling a constant vector of ones will always result in the same constant vector. Therefore, the permuted likelihood-ratio statistics will always be equal to the 'true' likelihood-ratio statistic, and the p-value will always be 1.

It may be possible to take a different approach. Just thinking out loud here for now - this will definitely not make it into permutes 2.3 (actually 2.3.2 now because of CRAN issues). It could be possible to use the Lee & Braun approach to partial out the random effects, and then instead of computing the fixed-effect permutations myself, I could have this done by lmPerm, which CAN permute the intercept (it uses a different approach based on orthogonal projections). I could then simply return those values. Might be worth looking into for v2.4...

**Comment by hy-lu** [comment]
Thank you for the explanation. I totally understand it and agree with you. Glad this issue has finally been worked out!

**Comment by cvoeten** [comment]
OK, glad we're on the same page now! I've released v2.3.2 now, which will make its way to CRAN shortly (where's v2.3 and v2.3.1, you ask? v2.3 I never released to CRAN pending resolution of this issue, v2.3.1 with these fixex did not pass CRAN checks, and while that issue did resolve itself automatically, it did necessitate a version bump.)
